### ISSUE TITLE
Interesting bug due to Swift's set indeterminism

### ISSUE THREAD

**Issue by zntfdr** [Iaenhaall/AttributedText#4]
TL;DR: it seems that Swift's set is causing a glitch in this library in `HTML2TextParser`.

# Problem 
Sometimes the _italics_ effect is not applied to the following text:  
`<b><u><i>bolded underlined and italized</i></u></b>, normal text`.

Reproduction code:
```swift
struct ContentView: View {
  var body: some View {
    List {
      ForEach(1..<100) { _ in
        AttributedText("<b><u><i>bolded underlined and italized</i></u></b>, normal text")
          .font(.system(size: 20))
      }
    }
  }
}
```

Result:
<img src="https://user-images.githubusercontent.com/5277837/171364017-f9aa12c3-5709-4819-b2e6-f6dc45bd419b.png" width="200">

Every time the app is launched, different rows will have the glitch above.

# Investigation

Initially, I thought it was a SwiftUI issue, however, this code works fine:

```swift
struct ContentView: View {
  var body: some View {
    List {
      ForEach(1..<100) { _ in
        Group {
          Text("bolded underlined and italized")
            .bold().italic().underline() // any sorting of this works fine in my tests.
          +
          Text(", normal text")
        }
        .font(.system(size: 20))
      }
    }
  }
}
```

After digging into the library code, I noticed that in `HTML2TextParser`'s `addChunkOfText(_:)` the following code is executed:

https://github.com/Iaenhaall/AttributedText/blob/26a8be0285d664449e352d0e9b1b4a79ee9cecba/Sources/AttributedText/HTML2TextParser.swift#L101-L105

...where `tags` is a `Set<String>`. 

Because Swift sets are unordered, every time the `for tag in tags` code is run, a new sequence of tags is produced and different parses of the same text (like the one above) will produce a different output.

In fact, if we replace the code `for tag in tags` with `for tag in tags.sorted()`, for every parse of the same text the same output will be produced. This is great, we've reached determinism. Except, now the glitch above is always present:
<img src="https://user-images.githubusercontent.com/5277837/171368719-1e93ecdb-4702-43ff-a215-b41db4211942.png" width="200">

The fix was to use another sorting order, this one works for this case: `for tag in tags.sorted(by: >)`:
<img src="https://user-images.githubusercontent.com/5277837/171369021-356d4ca7-9fcf-4dcc-a46b-534caf565a9f.png" width="200">

I'm not sure why having different tags orders cause the issue, if you have any idea, please let me know!

Thank you

**Comment by iaenhaall** [comment]
Thank you for the feedback and the research. I will try to investigate the problem and fix it as soon as I have some free time.

**Comment by LukasHakulin** [comment]
Hi guys!
I'm just looking for some simple html formatting solution with iOS 14 support (I'm new in swiftUI world) and I discover this library.
Unlikely I'm not able to use it for my purposes, but I spend a couple of hours playing with it and also checked known issues (just for sure :D).
I think what are you describing here isn't bug, it's just problem of your test, you are modifying results and this is the real reason why it's not working. 
Try that again, but "simpler":
```
struct ContentView: View {

    var body: some View {
        VStack(spacing: 20) {
            Group {
                Text("bold, italic, underline")
                    .fontWeight(.bold)
                    .italic()
                    .underline()
            }.background(Color.green)
            Group {
                Divider()
                Text("bold, italic, underline")
                    .bold()
                    .italic()
                    .underline()
                    .font(.system(size: 20))
                Text("bold, underline, italic")
                    .bold()
                    .underline()
                    .italic()
                    .font(.system(size: 20))
            }
            .padding(8)
            Group {
                Divider()
                Text("bold, italic, underline")
                    .fontWeight(.bold)
                    .italic()
                    .underline()
                    .font(.system(size: 20))
                Text("bold, underline, italic")
                    .fontWeight(.bold)
                    .underline()
                    .italic()
                    .font(.system(size: 20))
                Divider()
                Text("italic, bold, underline")
                    .italic()
                    .fontWeight(.bold)
                    .underline()
                    .font(.system(size: 20))
                Text("italic, underline, bold")
                    .italic()
                    .underline()
                    .fontWeight(.bold)
                    .font(.system(size: 20))
                Divider()
                Text("underline, bold, italic")
                    .underline()
                    .fontWeight(.bold)
                    .italic()
                    .font(.system(size: 20))
                Text("underline, italic, bold")
                    .underline()
                    .italic()
                    .fontWeight(.bold)
                    .font(.system(size: 20))
            }
        }
        .padding()
    }
}
```
![Screen Shot 2023-04-16 at 12 11 15](https://user-images.githubusercontent.com/8177839/232293960-1174c478-19ef-4f2a-a240-6aba71c559de.png)

From my point of view it looks like you have modified the result with application of Font on Group with tested data(Text) in combination with using `.bold()` instead of ` .fontWeight(.bold)` cause this is the real implementation of `<b>` tag in library.
Probably the easiest way to fix it is use `.bold()` as modifier.

Because ordering of modifiers matter, it's up to programmer to choose well to avoid formatting issues... or wait to iOS 15 and use AttributedString or some other solution ¯\_(ツ)_/¯
